File: programming_assignments/pa1/code/List.java
/* Adam Pinarbasi
   akpinarb
   pa1            */
import static java.lang.System.out;
import static java.lang.System.err;
import java.io.*;

class List {

   private class Node {
      int data;
      int index;
      Node next;
      Node prev;
      Node(int dataIn) {
         data  = dataIn;
         index = 0;
         next  = null;
         prev  = null;
      }

      // toString
      // Overriden toString method for class Node
      public String toString() {
         return String.valueOf(data);
      }
   }

   //********** Constructor **********

   List () {
      front  = null;
      back   = null;
      cursor = null;
      length = 0;
   }

   //********** Fields **********

   private Node front;
   private Node back;
   private Node cursor;
   private int length;

   //********** Access functions **********

   // length
   /* Returns the number of elements in this List */
   int length () { return length; }

   // index
   /* If cursor is defined, returns the index of the cursor element,
      otherwise returns -1 */
   int index () {
      return (cursor == null ? -1 : cursor.index);
   }

   // front
   /* Returns front element.  Pre: length() > 0 */
   int front () {
      if (length == 0) throw new RuntimeException
                       ("front(): length is 0\n");
      return front.data;
   }

   // back
   /* Returns back element. Pre: lenght() > 0 */
   int back () {
      if (length == 0) throw new RuntimeException
                       ("back(): length is 0\n");
      return back.data;
   }

   // get
   /* Returns element on cursor */
   int get () {
      if (cursor == null) throw new RuntimeException
                          ("cursorData(): cursor is null\n");
      return cursor.data;
   }

   // equals
   /* Returns true if this List and L are the same integer
      sequence. The cursor is ignored in both lists */
   boolean equals (List L) {
      if (L == null || length != L.length) return false;
      Node thisNode = front;
      for (Node LNode = L.front; LNode != null && thisNode != null; 
                LNode = LNode.next) {
         if (thisNode.data != LNode.data) return false;
         thisNode = thisNode.next;
      }
      return true;
   }

   //********** Manipulation procedures **********

   // clear
   /* Resets this List to its original empty state. */
   void clear () {
      front  = null;
      back   = null;
      cursor = null;
      length = 0;
   }

   // moveFront
   /* If List is non-empty, places the cursor under the front element, 
      otherwise does nothing */
   void moveFront () {
      if (length == 0) return;
      cursor = front;
   }

   // moveBack
   /* If List is non-empty, places the cursor under the back element,
      ootherwise does nothing */
   void moveBack () {
      if (length == 0) return;
      cursor = back;
   }

   // movePrev
   /* If cursor is defined and not at front, moves cursor one step toward
      front of this List, if cursor is defined and at front, cursor becomes
      undefined, if cursor is undefined does nothing */
   void movePrev () {
      if (cursor == null) return;
      if (cursor == front) { cursor = null; return; }
      cursor = cursor.prev;
   }


   // moveNext
   /* If cursor is defined and not at back, moves cursor one step toward
      back of this List, if cursor is defined and at back, cursor becomes
      undefined, if cursor is undefined does nothing */
   void moveNext () {
      if (cursor == null) return;
      if (cursor == back) { cursor = null; return; }
      cursor = cursor.next;
   }

   // prepend
   /* Insert new element into this List.  If List is non-empty,
      insertion takes place before front element */
   void prepend (int data) {
      Node prep = new Node(data);
      if (front == null) { front = prep; back = prep; }
      else {
         for (Node curr = front.next; curr != null; curr = curr.next) 
            ++curr.index;
         prep.next  = front;
         front.prev = prep;
         front      = prep;
      }
      ++length;
   }

   // append
   /* Insert new element into this List.  If List is non-empty, 
      insertion takes place after back element. */
   void append (int data) { 
      Node app = new Node(data);
      if (front == null) { 
         front = app;
         back  = app;
      }
      else {
         app.index = back.index + 1;
         app.prev  = back;
         back.next = app;
         back      = app;
      }
      ++length;
   }

   // insertBefore
   /* Insert new element before cursor.
      Pre: length() > 0, index() >= 0 */
   void insertBefore (int data) {
      if (cursor == null) throw new RuntimeException
                          ("insertBefore: cursor is null\n");
      Node insb  = new Node(data);
      insb.index = cursor.index;
      insb.next  = cursor;
      insb.prev  = cursor.prev;

      if (cursor.prev == null) front = insb;
      else cursor.prev.next = insb;

      cursor.prev = insb;
      for (Node curr = cursor; curr != null; curr = curr.next)
         ++curr.index;
      ++length;
   }

   // insertAfter
   /* Inserts new element after cursor.
      Pre: length() > 0, index() > = 0 */
   void insertAfter (int data) {
      if (cursor == null) throw new RuntimeException
                          ("insertAfter: cursor is null\n");
      Node insa  = new Node(data);
      insa.index = cursor.index;
      insa.prev  = cursor;
      insa.next  = cursor.next;

      if (cursor.next == null) back = insa; 
      else cursor.next.prev = insa;

      cursor.next = insa;
      for (Node curr = insa; curr != null; curr = curr.next)
         ++curr.index;
      ++length;
   }

   // deleteFront
   /* Deletes the front element. Pre: length() > 0 */
   void deleteFront () {
      if (length == 0) throw new RuntimeException
                       ("deleteFront: list is empty\n");
      front      = front.next;
      front.prev = null;
      for (Node curr = front; curr != null; curr = curr.next) 
         --curr.index; 
      --length;
   }

   // deleteBack
   /* Deletes the back element.  Pre: length() > 0 */
   void deleteBack () {
      if (length == 0) throw new RuntimeException
                       ("deleteBack: list is empty\n");
      back      = back.prev;
      back.next = null;
      --length;
   }

   // delete
   /* Deletes cursor element, making cursor undefined. 
      Pre: length() > 0, index() >= 0 */
   void delete () {
      if (cursor == null) throw new RuntimeException
                          ("delete: cursor is null\n");
      for (Node curr = cursor.next; curr != null; curr = curr.next) 
         --curr.index;
      cursor.prev.next = cursor.next;
      cursor.next.prev = cursor.prev;
      cursor = null;
      --length;
   }

   //********** Other methods **********

   // toString
   /* Overrides Object's toString method.  Returns a String
      representation of this List consisting of a space 
      separated sequence of integers, with front on left. 
      Returns null if list is empty */
   public String toString() {
      String list = front.toString(); 
      for (Node curr = front.next; curr != null; curr = curr.next) 
         list += " " + curr.toString();
      return list;
   }

   // copy
   /* Returns a new List representing the same integer sequence as this
      List.  The cursor in the new list is undefined, regardless of the 
      state of the cursor in this List.  This List is unchanged */
   List copy () {
      List copy = new List();  
      for (Node curr = front; curr != null; curr = curr.next) 
         copy.append(curr.data);
      return copy;
   }

   // concat
   /* Returns a new List which is the concatenation of 
      this list followed by L.  The cursor in the new List 
      is undefined, regardless of the states of the cursors
      in this List and L.  The states of this List and L are 
      unchanged */
   List concat (List L) {
      List cat = copy();
      for (Node curr = L.front; curr != null; curr = curr.next)
         cat.append(curr.data);
      return cat;
   }
}

File: programming_assignments/pa2/code/List.c
/* Adam Pinarbasi
   akpinarb
   pa2           */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include "List.h"

typedef struct node {
   struct node *next;
   struct node *prev;
   int data;
} node;

static void error (char *function, char *message) {
   printf("Error in function %s: %s\n", function, message);
   exit(EXIT_FAILURE);
}

// newList
/* List constructor */
List newList (void) {
   List thisList = malloc(sizeof(List));
   clear(thisList);
   return thisList;
}

node *newNode (void) {
   node *newNode = malloc(sizeof(node));
   newNode->next = NULL;
   newNode->prev = NULL;
   newNode->data = 0;
   return newNode;
}

// freeNode
/* node destructor */
void freeNode (node **dNode) {
   if (dNode != NULL && *dNode != NULL) {
      free(*dNode);
      *dNode = NULL;
   }
}

// freeList
/* List destructor */
void freeList (List *pL) {
   if (pL != NULL && *pL != NULL) {
      for (node *curr = (*pL)->front; curr != NULL;) {
         node *temp = curr;
         curr = curr->next;
         freeNode(&temp);
      }
      free(*pL); 
      *pL = NULL;
   }
}

// length
/* Get length */
int length (List L) { return L->length; }

// index
/* Get index */
int index (List L) { return (L->cursor == NULL ? -1 : L->index); }

// front
/* Get front element */
int front (List L) { 
   if (L->length == 0) error("front", "length is 0");
   return L->front->data; 
}

// back
/* Get back element */
int back (List L) { 
   if (L->back == NULL) error ("back", "back is NULL");
   return L->back->data; 
}

// get
/* Get element at cursor */
int get (List L) { 
   if (L->cursor == NULL) error("get", "cursor is NULL");
   return L->cursor->data; 
}

// equals
/* Find if List A equals List B */
int equals (List A, List B) {  
   if (A->length != B->length) return false;
   for (node *Anode = A->front, *Bnode = B->front; 
        Anode != NULL && Bnode != NULL;
        Anode = Anode->next, Bnode = Bnode->next ) 
      { if (Anode->data != Bnode->data) return false; }
   return true;
}
// clear
/* Clear out a List to defaults */
void clear (List L) { 
   L->front  = NULL;
   L->back   = NULL;
   L->cursor = NULL;
   L->index  = -1;
   L->length = 0;
}

// moveFront
/* Move cursor to front node */
void moveFront (List L) { 
   if (L->length == 0) error("moveFront", "length is 0");
   L->index = 0;
   L->cursor = L->front;
}

// moveBack
/* Move cursor to back node */
void moveBack (List L) {
   if (L->length == 0) error("moveBack", "length is 0"); 
  L->index = L->length - 1;
  L->cursor = L->back;
}

// movePrev
/* Move cursor to previous node */
void movePrev (List L) { 
   if (L->cursor == NULL) return;
   L->cursor = L->cursor->prev;
   --L->index;
}

// moveNext
/* Move cursor to next node */
void moveNext (List L) {
   if (L->cursor == NULL) return;
   L->cursor = L->cursor->next;
   ++L->index;
}

// prepend
/* Prepend node to list */
void prepend (List L, int data) { 
   node *prep    = newNode(); 
   prep->data    = data;
   prep->next    = L->front;
   if (L->front == NULL) {
      L->front   = prep;
      L->back    = prep;
   }
   else {
      L->front->prev = prep;
      L->front       = prep;
   }
   ++L->length;
}

// append
/* Append node to list */
void append (List L, int data) { 
   node *app     = newNode();
   app->data     = data;
   app->prev     = L->back;
   if (L->front == NULL) {
      L->front   = app;
      L->back    = app;
   }
   else {
      L->back->next = app;
      L->back       = app;
   }
   ++L->length;
}

// insertBefore
/* Insert node before cursor */
void insertBefore (List L, int data) {
   if (L->cursor == NULL) error("insertBefore", 
                               "cursor is NULL");
   node *insb = newNode();
   insb->data = data;
   insb->next = L->cursor;
   insb->prev = L->cursor->prev;
   if (L->cursor->prev != NULL) L->cursor->prev->next = insb;
   else L->front = insb;
   L->cursor->prev  = insb;
   ++L->index;
   ++L->length;
}

// insertAfter
/* Insert node after cursor */
void insertAfter (List L, int data) {
   if (L->cursor == NULL) error("insertBefore", 
                               "cursor is NULL");
   node *insa = newNode();
   insa->data = data;
   insa->prev = L->cursor;
   insa->next = L->cursor->next;
   if (L->cursor->next != NULL) L->cursor->next->prev = insa;
   else L->back = insa;
   L->cursor->next  = insa;
   ++L->length;
}

// deleteFront
/* Delete front element */
void deleteFront (List L) {
   if (L->front == NULL) error("deleteFront", 
                              "front is NULL");
   node *temp = L->front;
   L->front = L->front->next;
   free(temp);
   temp = NULL;
   --L->index;
}

// deleteBack
/* Delete back element */
void deleteBack (List L) {
   if (L->back == NULL) error("deleteBack", 
                             "back is NULL");
   node **temp = &L->back;
   L->back = L->back->prev;
   free(*temp);
   *temp = NULL;
   if (L->cursor == L->back) L->index = -1;
}

// delete
/* Delete element under cursor */
void delete (List L) {
   if (L->cursor == NULL) error("delete", 
                               "cursor is NULL");
   L->cursor->prev->next = L->cursor->next;
   L->cursor->next->prev = L->cursor->prev;
   free(L->cursor);
   L->cursor = NULL;
   //L->index = -1;
   --L->length;
}

// printList
/* Print the contents of the list */
void printList (FILE *out, List L) {
   for (node *curr = L->front; curr != NULL; curr = curr->next) {
      if (curr == L->back) fprintf(out, "%d", curr->data);
      else fprintf(out, "%d ", curr->data);
   }
}

// copyList
/* Return a new copy of the list */
List copyList (List L) {
   List newL = newList();
   for (node *curr = L->front; curr != NULL; curr = curr->next)
      append(newL, curr->data);
   return newL;
}

// concatList
/* Concatenate List B onto List A */
List concatList (List A, List B) {
   List cat = copyList(A);
   for (node *curr = B->front; curr != NULL; curr = curr->next)
      append(cat, curr->data);
   return cat;
}
